Add docs for sidecar injector watchdog (#2601)
dapr/docs: daprdocs/content/en/concepts/dapr-services/operator.md

## Running the operator service

The operator service is deployed as part of `dapr init -k`, or via the Dapr Helm charts. For more information on running Dapr on Kubernetes, visit the [Kubernetes hosting page]({{< ref kubernetes >}}).

## Additional configuration options

The operator service includes additional configuration options.

### Injector watchdog

> The operator service includes an _injector watchdog_ feature which periodically polls all pods running in your Kubernetes cluster and confirms that the Dapr sidecar is injected in those which have the `dapr.io/enabled=true` annotation. It is primarily meant to address situations where the [sidecar injector service]({{< ref sidecar-injector >}}) did not successfully inject the sidecar (the `daprd` container) into pods.

Reviewer: ```suggestion
The operator service includes an _injector watchdog_ feature which periodically polls all pods running in your Kubernetes cluster and confirms that the Dapr sidecar is injected in those which have the `dapr.io/enabled=true` annotation. It is primarily meant to address situations where the [Injector service]({{< ref sidecar-injector >}}) did not successfully inject the sidecar (the `daprd` container) into pods.

```

The injector watchdog can be useful in a few situations, including:

> - Recovering from a Kubernetes cluster completely stopped. When a cluster is completely stopped and then restarted (including in the case of a total cluster failure), pods are restarted in a random order. If your application is restarted before the Dapr control plane (specifically the sidecar injector service) is ready, the Dapr sidecar may not be injected into your application's pods, causing your application to behave unexpectedly.

Reviewer: ```suggestion
- Recovering from a Kubernetes cluster completely stopped. When a cluster is completely stopped and then restarted (including in the case of a total cluster failure), pods are restarted in a random order. If your application is restarted before the Dapr control plane (specifically the Injector service) is ready, the Dapr sidecar may not be injected into your application's pods, causing your application to behave unexpectedly.

```

## Additional configuration options

The operator service includes additional configuration options.

### Injector watchdog

The operator service includes an _injector watchdog_ feature which periodically polls all pods running in your Kubernetes cluster and confirms that the Dapr sidecar is injected in those which have the `dapr.io/enabled=true` annotation. It is primarily meant to address situations where the [Injector service]({{< ref sidecar-injector >}}) did not successfully inject the sidecar (the `daprd` container) into pods.

The injector watchdog can be useful in a few situations, including:

- Recovering from a Kubernetes cluster completely stopped. When a cluster is completely stopped and then restarted (including in the case of a total cluster failure), pods are restarted in a random order. If your application is restarted before the Dapr control plane (specifically the Injector service) is ready, the Dapr sidecar may not be injected into your application's pods, causing your application to behave unexpectedly.

> - Addressing potential random failures with the sidecar injector, such as transient failures within the sidecar injector service.

Reviewer: ```suggestion
- Addressing potential random failures with the sidecar injector, such as transient failures within the Injector service.

```

If the watchdog detects a pod that does not have a sidecar when it should have had one, it deletes it. Kubernetes will then re-create the pod, invoking the Dapr sidecar injector again.

> The injector watchdog feature is **disabled by default**.

Reviewer: ```suggestion
The injector watchdog is **disabled by default**.

```

Reviewer: Why is this disabled by default? Can this not be enabled by default?

Author: Being a new thing, @artursouza recommended it to be disabled by default. We also don't really know if we can set a default interval for everyone:

- An interval that's too small will increase the load on the K8s API server possibly by a lot
- An interval that's too big could make the RTO (Recovery Time Objective) too big for customers

> You can enable it by passing the `--watch-interval` flag to the `operator` command, which can assume one of the following values:

Reviewer: ```suggestion
You can enable it by passing the `--watch-interval` flag to the `operator` command, which can take one of the following values:

```

Author: I don't think this link is the correct one. The link goes to the Helm options but this paragraph is about the CLI command.

Reviewer: Removed link

> You can enable it by passing the `--watch-interval` flag to the `operator` command, which can assume one of the following values:

Reviewer: So the only way I can do this realistically is via Helm?

Author: Assuming you're using Helm, yes.

Otherwise, if you're running Dapr on Kubernetes with your own chart (not sure if anyone does that, but...), then you need to pass the flag to the command

- `--watch-interval=0`: disables the injector watchdog (default value if the flag is omitted).
- `--watch-interval=<interval>`: the injector watchdog is enabled and polls all pods at the given interval; the value for the interval is a string that includes the unit. For example: `--watch-interval=10s` (every 10 seconds) or `--watch-interval=2m` (every 2 minutes).
- `--watch-interval=once`: the injector watchdog runs only once when the operator service is started.

> If you're using Helm, you can configure the injector watchdog with the `dapr_operator.watchInterval` option, which can have the same values as the command line flag.

Reviewer: ```suggestion
If you're using Helm, you can configure the injector watchdog with the [`dapr_operator.watchInterval` option](https://github.com/dapr/dapr/blob/master/charts/dapr/README.md#dapr-operator-options), which has the same values as the command line flags.

```

> > The injector watchdog is safe to use when the operator service is running in HA (High Availability) mode with more than one replica. In this case, Kubernetes automatically elects a "leader" instance which is the only one that will run the injector watchdog service.  

Reviewer: ```suggestion
> The injector watchdog is safe to use when the operator service is running in HA (High Availability) mode with more than one replica. In this case, Kubernetes automatically elects a "leader" instance which is the only one that runs the injector watchdog service.  

```

> > The injector watchdog is safe to use when the operator service is running in HA (High Availability) mode with more than one replica. In this case, Kubernetes automatically elects a "leader" instance which is the only one that will run the injector watchdog service.  

Reviewer: Suggest adding a link to your video here. The YouTube URL reference below is correct. Add this below

Watch this video for an overview of the watchdog injector.

<div class="embed-responsive embed-responsive-16by9">
<iframe width="360" height="315" src="https://youtu.be/ecFvpp24lpo?t=1848" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
